File: PowerXN.py
#TC: O(logN)
#SC: O(logN)

class Solution:
    def myPow(self, x: float, n: int) -> float:

        if (n < 0):
            if (x == 0):             
                return float('inf')
            x = 1 / x
            n = -1 * n

        if (n == 0):
            return 1
        if (n == 1):
            return x

        val = self.myPow(x, int(n / 2))

        if (n % 2 == 0):
            return val * val
        else:
            return x * val * val

File: PowerXN.cs
using System;
namespace Algorithms
{
    public class PowerXN
    {
        /// Time Complexity : O(Logn)
        // Space Complexity :O(n)
        // Did this code successfully run on Leetcode :Yes
        // Any problem you faced while coding this :  No
        public double MyPow(double x, int n)
        {
            //base
            if (n == 0) return 1.0000;
            double result = MyPow(x, n / 2);
            if (n % 2 == 0)
            {
                result = result * result;
            }
            else
            {
                if (n < 0)
                {
                    result = result * result * 1 / x;
                }
                else
                {
                    result = result * result * x;
                }
            }
            return result;
        }
    }
}
